### PR TITLE
EZP-31383: Implemented Role copying API

### PR DESCRIPTION
**Pull request by barw4**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [EZP-31383](https://jira.ez.no/browse/EZP-31383)
| **Original PR**                        | ezsystems/ezpublish-kernel#3009
| **Type**                                   | feature
| **Target eZ Platform version** | `ezplatform-kernel:1.0`
| **BC breaks**                          | no
| **Tests pass**                          | yes
| **Doc needed**                       | no

Prepared `ezplatform-kernel` for `Role` copying functionality.
Added `copyRole` method to `RoleService` API Interface with `Core` and `Event` implementations.
Created related structs and `Database Gateway` and `User\Handler` implementations.
Added `Role` copying integration tests using `RoleService` API with double-check on`Policies` and `Limitations` respectively.

Related `admin-ui` PR: https://github.com/ezsystems/ezplatform-admin-ui/pull/1241

#### Checklist:
- [x] PR description is updated.
- [x] Tests are implemented.
- [x] Added code follows Coding Standards (use `$ composer fix-cs`).
- [x] PR is ready for a review.

### CODE REVIEW

File: eZ/Publish/SPI/Persistence/User/RoleCopyStruct.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     * Status of new role.
+     *
+     * @var string
+     */
+    public $status;
+
+    /**
+     * Contains an array of role policies.
+     *
+     * @var array
```
Comment (adamwojs) [inline eZ/Publish/SPI/Persistence/User/RoleCopyStruct.php]: Is it possible to specify more  accurate type information ?

File: eZ/Publish/API/Repository/Values/User/Role.php [hunk 2]
```diff
@@ -48,6 +48,14 @@ abstract class Role extends ValueObject
      */
     protected $status;
 
+    /**
+     * @return int
+     */
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/User/Role.php]:
```suggestion
```

File: eZ/Publish/Core/Persistence/Cache/UserHandler.php [hunk 3]
```diff
@@ -275,6 +276,16 @@ public function createRoleDraft($roleId)
         return $this->persistenceHandler->userHandler()->createRoleDraft($roleId);
     }
 
+    /**
+     * {@inheritdoc}
+     */
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Cache/UserHandler.php]:
```suggestion
```

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 4]
```diff
@@ -86,6 +86,36 @@ public function createRole(Role $role): Role
...
+            ->insert(self::ROLE_TABLE)
+            ->values(
+                [
+                    'is_new' => $query->createPositionalParameter(0, ParameterType::INTEGER),
+                    'name' => $query->createPositionalParameter(
+                        $role->identifier,
+                        ParameterType::STRING
+                    ),
+                    'value' => $query->createPositionalParameter(0, ParameterType::INTEGER),
+                    // BC: "version" stores originalId when creating a draft from an existing role
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: Why we need to store original role id ?

**Comment by barw4** [comment]
@adamwojs resolved in: https://github.com/ezsystems/ezplatform-kernel/pull/16/commits/6bf5e22432f9c9f3d065ae2b5ab1e6049f448ec8

**Pull request merged**